### PR TITLE
RA2-231-COG-analysis

### PR DESCRIPTION
**Pull request by Bluesmile82**
Analysis COGs and categorical layers
 
image

image

- Adapt the fetch for COGs layers
- Display Categorical layers on a new horizontal bar chart
- Display Continuous layers on the existing histogram chart

## Testing instructions

- Go to map
1.
- Select Land Map test layers => Land Cover layer
- Click the Analysis button
- Analyze some area drawing or selecting a country
- You should see the chart and info

2.
- Select Land Map test layers =>wheat [continous cog]
- Click the Analysis button
- Analyze some area drawing or selecting a country (better on the Horn of Africa)
- You should see the chart and info

Attention: The download service is not working. This is happening also in production and staging

## Tracking

https://vizzuality.atlassian.net/browse/RA2-126?atlOrigin=eyJpIjoiZTY1ZDNlZGIzMTMyNDg5N2I2N2NmY2YwM2I1Y2FhOTMiLCJwIjoiaiJ9

https://vizzuality.atlassian.net/browse/RA2-16?atlOrigin=eyJpIjoiNGNiYmVlOGVhMGFjNDA5YTk0MmI2NGJmMmE1NmZhYjciLCJwIjoiaiJ9

https://vizzuality.atlassian.net/browse/RA2-227?atlOrigin=eyJpIjoiZThmYzdkMjk4ZGRhNDNjMWI2M2VjYzkzNGM4NzAwNjEiLCJwIjoiaiJ9

### CODE REVIEW

**Comment by Bluesmile82** [comment]
@clementprdhomme I have also added the continuous layers analysis for you to review when you are available

File: frontend/src/utilities/hooks/useWidget.js [hunk 1]
```diff
@@ -4,17 +4,29 @@ import { useAxios } from './useAxios';
...
+export const useWidget = ({ slug, geojson }, { type, analysisQuery, analysisBody }) => {
+  const isCOGLayer = useMemo(() => type === 'cog', [type]);
   const query = useMemo(() => {
     if (analysisBody) {
       const { assetId } = JSON.parse(analysisBody);
+      let parsedQuery = analysisQuery;
+
+      if (isCOGLayer) {
+        const parsedBody = JSON.parse(analysisBody);
+        const { params } = parsedBody || {};
```
Comment (clementprdhomme) [inline frontend/src/utilities/hooks/useWidget.js:16]: You can extract `params` from line 11 😉.

File: frontend/src/utilities/hooks/useWidget.js [hunk 2]
```diff
@@ -4,17 +4,29 @@ import { useAxios } from './useAxios';
...
+        Object.entries(params).forEach(([key, value]) => {
```
Comment (clementprdhomme) [inline frontend/src/utilities/hooks/useWidget.js:17]: Don't you need to check if `params` is an object?

File: frontend/src/views/shared/Widgets/WidgetCategoricalChart.jsx [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+      <foreignObject y={y} width={chartWidth} height={16}>
+        <div className="categorical-labels with-ellipsis">{name}</div>
+      </foreignObject>
+      <Rectangle
+        {...props}
+        fill={color}
+        y={y + barHeight / 2 + RECTANGLE_PADDING}
+        height={barHeight - barHeight / 2}
+      />
+      ;
```
Comment (clementprdhomme) [inline frontend/src/views/shared/Widgets/WidgetCategoricalChart.jsx:34]: I think this line was unintended, right?

File: frontend/src/views/shared/Widgets/WidgetBarChart.jsx [hunk 4]
```diff
@@ -23,13 +23,33 @@ export const WidgetBarChart = ({
...
+          .sort((a, b) => a.min - b.min)
+      : data?.rows;
+  }, [data, isCOG]);
+
+  const downloadData = useMemo(() => {
+    return isCOG
+      ? {
+          fields: ['name', 'count'],
+          rows: mergedBarData?.map((d) => ({ name: d.name, min: d.count })),
+        }
```
Comment (clementprdhomme) [inline frontend/src/views/shared/Widgets/WidgetBarChart.jsx:50]: I think there's an issue with the CSV. Everytime I downloaded it, the structure was like this one:
```csv
0,1
,5333
,1
,13
,4
,2051
,0
,0
,3
,0
,1
```

Here's the chart
![Analysis of Wheat [continuous cog]](https://github.com/ConservationInternational/resilienceatlas/assets/6073968/653ebd3d-75b4-4e64-9bf7-722fe97fe8e2)

File: frontend/src/views/shared/Widgets/WidgetCategoricalChart.jsx [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+                </li>
+                <li className="stats-item">
+                  <span className="stats-item__label">
+                    <T _str="Sum:" _comment="Sum: on the widget bar chart" />{' '}
+                  </span>
+                  {formatNumber({ value: data.stats.sum })}&nbsp;
+                  {unit}
+                </li>
+              </ul>
+            )}
```
Comment (clementprdhomme) [inline frontend/src/views/shared/Widgets/WidgetCategoricalChart.jsx:161]: I'm not sure these stats make sense for this chart and there are not part of the design so I think we could remove them.

File: frontend/src/views/shared/Widgets/WidgetCategoricalChart.jsx [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+        return { count: d.count, color: legendItem?.value, name: legendItem?.name };
+      })
+      .sort((a, b) => b.count - a.count);
+  }, [data, legendData]);
+  const downloadData = useMemo(() => {
+    return {
+      fields: ['name', 'count'],
+      rows: mergedBarData?.map((d) => ({ name: d.name, count: d.count })),
+    };
+  }, [mergedBarData]);
```
Comment (clementprdhomme) [inline frontend/src/views/shared/Widgets/WidgetCategoricalChart.jsx:72]: The CSV is incorrect. You need to escape the commas “,” from the labels 😉.

**Comment by Bluesmile82** [comment]
Ready again! The CSVs have been fixed and your comments addressed. Thanks!

File: frontend/src/utilities/hooks/useWidget.js [hunk 7]
```diff
@@ -4,17 +4,28 @@ import { useAxios } from './useAxios';
...
+export const useWidget = ({ slug, geojson }, { type, analysisQuery, analysisBody }) => {
+  const isCOGLayer = useMemo(() => type === 'cog', [type]);
   const query = useMemo(() => {
     if (analysisBody) {
-      const { assetId } = JSON.parse(analysisBody);
+      const { assetId, params } = JSON.parse(analysisBody);
+      let parsedQuery = analysisQuery;
+
+      if (isCOGLayer && params && typeof params === 'object') {
+        params &&
```
Comment (clementprdhomme) [inline frontend/src/utilities/hooks/useWidget.js:15]: I think this line is not needed since you're already checking if `params` exists above.

**Review by clementprdhomme — APPROVED** [review]

**Pull request merged**